### PR TITLE
Remove deprecated SubDags

### PR DESCRIPTION
**Pull request by kaxil**
This PR removes SubDags in favor of TaskGroups fro Airflow 3.0

Subdags have been removed from the following locations:

- CLI
- API
- ``SubDagOperator``

This removal marks the end of Subdag support across all interfaces. Users should transition to using TaskGroups as a more efficient and maintainable alternative.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/dataplex.py [hunk 1]
```diff
@@ -279,14 +279,7 @@ def get_task(
...
         name = f"projects/{project_id}/locations/{region}/lakes/{lake_id}/tasks/{dataplex_task_id}"
         client = self.get_dataplex_client()
-        result = client.get_task(
-            request={
-                "name": name,
-            },
-            retry=retry,
-            timeout=timeout,
-            metadata=metadata,
-        )
```
Comment (kaxil) [inline airflow/providers/google/cloud/hooks/dataplex.py]: hmm unrelated change

File: airflow/providers/fab/auth_manager/security_manager/override.py [hunk 2]
```diff
@@ -1073,15 +1073,15 @@ def create_dag_specific_permissions(self) -> None:
         dags = dagbag.dags.values()
 
         for dag in dags:
-            root_dag_id = dag.parent_dag.dag_id if dag.parent_dag else dag.dag_id
```
Comment (kaxil) [inline airflow/providers/fab/auth_manager/security_manager/override.py:1076]: We might need to keep this since this is in FAB provider

File: airflow/api/common/delete_dag.py [hunk 3]
```diff
@@ -64,18 +64,6 @@ def delete_dag(dag_id: str, keep_records_in_log: bool = True, session: Session =
...
-    dags_to_delete_query = session.execute(
-        select(DagModel.dag_id).where(
-            or_(
-                DagModel.dag_id == dag_id,
-                and_(DagModel.dag_id.like(f"{dag_id}.%"), DagModel.is_subdag),
-            )
-        )
-    )
-
-    dags_to_delete = [dag_id for (dag_id,) in dags_to_delete_query]
```
Comment (uranusjr) [inline airflow/api/common/delete_dag.py:77]: This would fail to actually delete the non-sub DAG (the `== dag_id` part)

**Comment by eladkal** [comment]
We also have the UI part of Subdag (The zoom in button)

https://github.com/apache/airflow/blob/1074b8eed680af9668f11c63cf28e72db5470fde/airflow/www/static/js/dag/details/taskInstance/Nav.tsx#L70-L85

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 4]
```diff
@@ -495,8 +495,9 @@ def _get_ti_pod_labels(context: Context | None = None, include_try_number: bool
         if include_try_number:
             labels.update(try_number=ti.try_number)
         # In the case of sub dags this is just useful
-        if context["dag"].parent_dag:
+        if hasattr(context["dag"], "parent_dag") and context["dag"].parent_dag:
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/operators/pod.py]:
```suggestion
        if getattr(context["dag"], "parent_dag", False):
```

File: airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py [hunk 5]
```diff
@@ -202,7 +202,7 @@ def create_labels_for_pod(context: dict | None = None, include_try_number: bool
             labels.update(try_number=ti.try_number)
 
         # In the case of sub dags this is just useful
-        if context["dag"].is_subdag:
+        if hasattr(context["dag"], "is_subdag") and context["dag"].is_subdag:
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py]:
```suggestion
        if getattr(context["dag"], "is_subdag", False):
```

File: airflow/providers/fab/auth_manager/security_manager/override.py [hunk 6]
```diff
@@ -1073,7 +1073,10 @@ def create_dag_specific_permissions(self) -> None:
...
+            if hasattr(dag, "parent_dag") and dag.parent_dag:
+                root_dag_id = dag.parent_dag.dag_id
+            else:
+                root_dag_id = dag.dag_id
```
Comment (uranusjr) [inline airflow/providers/fab/auth_manager/security_manager/override.py]:
```suggestion
            root_dag_id = (getattr(dag, "parent_dag", None) or dag).dag_id
```

File: tests/www/views/test_views_tasks.py [hunk 7]
```diff
@@ -976,7 +957,7 @@ def test_task_instance_set_state_failure(admin_client, action):
     "task_search_tuples",
     [
         [("example_xcom", "bash_push"), ("example_bash_operator", "run_this_last")],
-        [("example_subdag_operator", "some-other-task")],
+        [("example_python_operator", "print_the_context")],
```
Comment (uranusjr) [inline tests/www/views/test_views_tasks.py]: I think we can get rid of this case entirely? It would just test the same things as the first one if the operator not a subdag one.

File: airflow/api/common/delete_dag.py (hunk 3 above)
Comment (kaxil) [reply airflow/api/common/delete_dag.py:77]: I am handling it at the bottom -- Line 77

File: tests/api_experimental/common/test_mark_tasks.py [hunk 8]
```diff
@@ -769,8 +742,6 @@ def test_set_state_with_multiple_dagruns(self, session=None):
         # Recursively count number of tasks in the dag
         def count_dag_tasks(dag):
             count = len(dag.tasks)
-            subdag_counts = [count_dag_tasks(subdag) for subdag in dag.subdags]
-            count += sum(subdag_counts)
             return count
 
         assert len(altered) == count_dag_tasks(self.dag2)
```
Comment (uranusjr) [inline tests/api_experimental/common/test_mark_tasks.py]: This should just be `assert len(altered) == len(self.dag2.tasks)`

File: airflow/www/views.py [hunk 9]
```diff
@@ -844,7 +844,7 @@ def index(self):
 
         with create_session() as session:
             # read orm_dags from the db
-            dags_query = select(DagModel).where(~DagModel.is_subdag, DagModel.is_active)
+            dags_query = select(DagModel).where(DagModel.is_active)
```
Comment (uranusjr) [inline airflow/www/views.py:847]: This combined with the model change means all previously subDAGs will now show up in the index page as normal DAGs. Should we provide a way for users to remove all those subDAGs?

File: newsfragments/41390.significant.rst
Comment (uranusjr) [inline newsfragments/41390.significant.rst:1]: Not specific to this PR but for all cleanup PRs… where should we write up a more detailed description on what breaks and what should users to do migrate? (For example, this PR means all subDAGs become normal DAGs, and many REST API fields are gone.)

The release notes page does not seem to be a good place (it’s be way too long) but we should write the document when we make the changes so things don’t get forgotten when the information is needed (user will want it).

Comment (kaxil) [reply newsfragments/41390.significant.rst:1]: Currently, we are capturing it in newsfragment but I think your point is it might not be sufficient? For 2.0.0, we saved all those details in [`Updating.md`](https://github.com/apache/airflow/blob/2.0.1/UPDATING.md#airflow-200) & [`upgrading-to-2.rst`](https://github.com/apache/airflow/blob/2.0.1/docs/apache-airflow/upgrading-to-2.rst) -- rendered as : https://airflow.apache.org/docs/apache-airflow/2.0.0/upgrading-to-2.html

@jedcunningham / @ephraimbuddy / @eladkal Any thoughts on this?

File: airflow/www/views.py (hunk 9 above)
Comment (kaxil) [reply airflow/www/views.py:847]: Maybe this is where we use the "Upgrade/Migration tool" too. For just core Airflow perspective, maybe we can add a migration to remove those from `dag` table?

File: newsfragments/41390.significant.rst
Comment (jedcunningham) [reply newsfragments/41390.significant.rst:1]: My take is newfragments are a great way to keep tabs on it for now, and we will write more/better docs at a later date, once we have a more complete picture of everything.

Comment (potiuk) [reply newsfragments/41390.significant.rst:1]: +1 on that one. We might find out during the process that some of the changes will be combined, have several steps, they can be withdrawn etc, and we should just "capture" what's been updated now and produce more details afterwards. There are a number of semi-overlaping changes coming and it will be difficult to keep this document up-to-date all the time. Also such document should be consistently prepared with good structure. Possibly it should be written by a technical writer with input of all the contributors who added breaking changes.

Comment (eladkal) [reply newsfragments/41390.significant.rst:1]: I agree

File: airflow/www/views.py (hunk 9 above)
Comment (uranusjr) [reply airflow/www/views.py:847]: We could but that would require this PR to not remove the db field (just yet), otherwise it would be impossible to tell what DAGs should be removed.

File: tests/conftest.py [hunk 10]
```diff
@@ -847,6 +847,12 @@ def get_serialized_data(self):
                 return json.loads(data)
             return data
 
+        def _bag_dag_compat(self, dag):
```
Comment (uranusjr) [inline tests/conftest.py:847]: Why do we need this? Isn’t the test always run in newer versions? Or is this supposed to be removed when/before we release 3.0 or something?

Comment (kaxil) [reply tests/conftest.py:847]: Because of our directory structure, tests in providers inherit this conftest.

Comment (uranusjr) [reply tests/conftest.py:847]: I see. Maybe a note here for this and a todo to remove this when all providers drop 2.x support?

File: airflow/www/views.py (hunk 9 above)
Comment (kaxil) [reply airflow/www/views.py:847]: The Migration Utilities could be run before upgrading to 3.x -- so the environment would be ready for 3.0. 

image

Comment (kaxil) [reply airflow/www/views.py:847]: (Similar to how we had it for 2.0 -- Step 3 in the above figure is where we ran it whereas we upgrade on Step 7)

https://airflow.apache.org/docs/apache-airflow/2.0.0/upgrading-to-2.html

Comment (kaxil) [reply airflow/www/views.py:847]: I could also delete those records in the DB migration if we'd like that

Comment (uranusjr) [reply airflow/www/views.py:847]: Unconditionally deleting all subDAGs is a bit too heavy-handed to me. We could do that _if_ the migration tool allows the user to mark what DAGs to not be removed (maybe by setting their `is_subdag` flag to False somehow?)

Comment (kaxil) [reply airflow/www/views.py:847]: In that case, for now, we should rely on the migration tool to delete or change it to main DAG IMO.

File: tests/conftest.py (hunk 10 above)
Comment (kaxil) [reply tests/conftest.py:847]: Added

File: airflow/api/common/delete_dag.py (hunk 3 above)
Comment (ephraimbuddy) [reply airflow/api/common/delete_dag.py:77]: Does this not remove deletion altogether?

Comment (ephraimbuddy) [reply airflow/api/common/delete_dag.py:77]: seen the line 77 :)

**Review by ephraimbuddy — COMMENTED** [review]
Concerning the root_dag_id, I guess `security/permissions.py` should also be updated but then, I think Airflow 3 might no longer be compatible with current providers.

**Review by phanikumv — COMMENTED** [review]
We'll also need to remove docs/apache-airflow/img/subdag_after.png ,  docs/apache-airflow/img/subdag_before.png and docs/apache-airflow/img/subdag_zoom.png

**Comment by kaxil** [comment]
> We'll also need to remove docs/apache-airflow/img/subdag_after.png , docs/apache-airflow/img/subdag_before.png and docs/apache-airflow/img/subdag_zoom.png

Done, removed

**Comment by jscheffl** [comment]
Oh, in numer of LoC removed as PR this is a high chance to get PR of the month :-D

File: airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+def upgrade():
+    """Remove ``is_subdag`` column from DAGs table."""
+    conn = op.get_bind()
+    inspector = sa.inspect(conn)
+
+    with op.batch_alter_table("dag", schema=None) as batch_op:
+        if _index_exists(inspector, "idx_root_dag_id"):
+            batch_op.drop_index("idx_root_dag_id")
+        if _column_exists(inspector, "is_subdag"):
```
Comment (jedcunningham) [inline airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py:59]: Why are we doing these conditionally? How could the columns not exist?

Comment (kaxil) [reply airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py:59]: I didn't realize most of our migration aren't idempotent. I had a first pass at those few years back : https://github.com/apache/airflow/pulls?q=is%3Apr+idempotent+migration+is%3Aclosed+author%3Akaxil

This is great for cases where downgrades or upgrades don't complete fully. You can rerun run.

Comment (jedcunningham) [reply airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py:59]: I see. That's really only an issue on our old friend mysql, who lacks transactional DDL 🙃.

**Review by jedcunningham — APPROVED** [review]

**Review by potiuk — COMMENTED** [review]
So much complexity removed ... Good job.

File: airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py (hunk 11 above)
Comment (jedcunningham) [reply airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py:59]: You could make the argument that, in particular when adding, this "if not exists" idempotency is actually opening up the possibility for inconsistency. What if the column type changed later? Probably better to let it fail if the db is in an inconsistent state.

**Comment by potiuk** [comment]
BTW. One more comment https://github.com/apache/airflow/issues/21867 -> I believe retry for failed task group is the only serious missing feature for Task Groups comparing to SubDags. 

https://github.com/apache/airflow/issues/21867

We should likely consider implementing it.

File: airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py (hunk 11 above)
Comment (kaxil) [reply airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py:59]: Column type shouldn't matter though since we are just checking if the column exists or not in that condition -- rest all remains same.

**Comment by kaxil** [comment]
> BTW. One more comment https://github.com/apache/airflow/issues/21867 -> I believe retry for failed task group is the only serious missing feature for Task Groups comparing to SubDags. 
> 
> 
> 
> Retry for Task Groups: https://github.com/apache/airflow/issues/21867
> 
> 
> 
> We should likely consider implementing it.

Yup, added a label to that issue to track it

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 12]
```diff
@@ -495,8 +495,9 @@ def _get_ti_pod_labels(context: Context | None = None, include_try_number: bool
         if include_try_number:
             labels.update(try_number=ti.try_number)
         # In the case of sub dags this is just useful
-        if context["dag"].parent_dag:
-            labels["parent_dag_id"] = context["dag"].parent_dag.dag_id
+        if getattr(context["dag"], "parent_dag", False):
+            labels["parent_dag_id"] = context["dag"].parent_dag.dag_id  # type: ignore[attr-defined]
```
Comment (eladkal) [inline airflow/providers/cncf/kubernetes/operators/pod.py:499]: @kaxil What does this change means for users running newer provider versions with Airflow 2?
Isn't that breaking change? labeling will not work as expected for SubDAGs?

Comment (potiuk) [reply airflow/providers/cncf/kubernetes/operators/pod.py:499]: from what I see - they wil ljust miss "parent_dag_id" label, as getattr wll handle lack of parent_dag field - and yes, it's a breaking change for those who relied on subdags, but that's expected in this case.

Comment (eladkal) [reply airflow/providers/cncf/kubernetes/operators/pod.py:499]: but we can avoid it don't we?
we can place this change under check version flag

Comment (potiuk) [reply airflow/providers/cncf/kubernetes/operators/pod.py:499]: We don't have to. It happens automatically via getattr - it will work in backwards-compatible way if the context will contain parent_dag automatically.

Comment (potiuk) [reply airflow/providers/cncf/kubernetes/operators/pod.py:499]: But you are quite right - we should  (@kaxil?) likely add some flag or at least comment to mark it as removable when we drop Airflow 2 compatibility for providers.

Comment (kaxil) [reply airflow/providers/cncf/kubernetes/operators/pod.py:499]: >We don't have to. It happens automatically via getattr - it will work in backwards-compatible way if the context will contain parent_dag automatically.

Precisely, this check is backwards-compatible. For someone using sub-dags, it works same as before.

I'll add a TODO in a separate PR to remove this check when we drop Airflow 2 support in Providers.

Comment (kaxil) [reply airflow/providers/cncf/kubernetes/operators/pod.py:499]: @eladkal: I created a PR to address this: https://github.com/apache/airflow/pull/41963
